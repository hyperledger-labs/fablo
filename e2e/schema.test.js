import { matchers } from "jest-json-schema";
import * as schema from "../docs/schema.json";
import * as docsSample from "../docs/sample.json";
import { FabloConfigJson, PrivateDataJson } from "../src/types/FabloConfigJson";

expect.extend(matchers);

describe("schema", () => {
  const base = docsSample;
  expect(base).toMatchSchema(schema);

  const updatedBase = (updateJson: (_: FabloConfigJson) => void): Record<string, unknown> => {
    const json = JSON.parse(JSON.stringify(base));
    updateJson(json as FabloConfigJson);
    return json;
  };

  const lettersOnly = "lettersonly";
  const underscore = "under_score";
  const lettersAndNumber = "lettersand4";
  const uppercase = "UpperCase";
  const domain = "domain.example.com";
  const spaces = "with spaces";
  const specialCharacters1 = "with!characters";
  const specialCharacters2 = "withspecial@";

  it("should match snapshot", () => {
    expect(schema).toMatchSnapshot();
  });

  it("should validate fabric version", () => {
    const withFabricVersion = (v: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.networkSettings.fabricVersion = v;
      });

    expect(withFabricVersion("1.4.3")).toMatchSchema(schema);
    expect(withFabricVersion("1.4.4")).toMatchSchema(schema);
    expect(withFabricVersion("1.3.1")).not.toMatchSchema(schema);
    expect(withFabricVersion("2.0.0")).not.toMatchSchema(schema);
  });

  // it("should validate root org name", () => {
  //   const withRootOrgName = (n: string) =>
  //     updatedBase((json: FabloConfigJson) => {
  //       json.rootOrg.organization.name = n;
  //     });
  //
  //   expect(withRootOrgName(lettersOnly)).toMatchSchema(schema);
  //   expect(withRootOrgName(lettersAndNumber)).toMatchSchema(schema);
  //   expect(withRootOrgName(uppercase)).toMatchSchema(schema);
  //   expect(withRootOrgName(domain)).not.toMatchSchema(schema);
  //   expect(withRootOrgName(spaces)).not.toMatchSchema(schema);
  //   expect(withRootOrgName(specialCharacters1)).not.toMatchSchema(schema);
  //   expect(withRootOrgName(specialCharacters2)).not.toMatchSchema(schema);
  // });
  //
  // it("should validate root org MSP name", () => {
  //   const withRootOrgMSPName = (n: string) =>
  //     updatedBase((json: FabloConfigJson) => {
  //       json.rootOrg.organization.mspName = n;
  //     });
  //
  //   expect(withRootOrgMSPName(lettersOnly)).toMatchSchema(schema);
  //   expect(withRootOrgMSPName(lettersAndNumber)).toMatchSchema(schema);
  //   expect(withRootOrgMSPName(uppercase)).toMatchSchema(schema);
  //   expect(withRootOrgMSPName(domain)).not.toMatchSchema(schema);
  //   expect(withRootOrgMSPName(spaces)).not.toMatchSchema(schema);
  //   expect(withRootOrgMSPName(specialCharacters1)).not.toMatchSchema(schema);
  //   expect(withRootOrgMSPName(specialCharacters2)).not.toMatchSchema(schema);
  // });

  // it("should validate root org domain", () => {
  //   const withRootOrgDomain = (d: string) =>
  //     updatedBase((json: FabloConfigJson) => {
  //       json.rootOrg.organization.domain = d;
  //     });
  //
  //   expect(withRootOrgDomain(lettersOnly)).toMatchSchema(schema);
  //   expect(withRootOrgDomain(lettersAndNumber)).toMatchSchema(schema);
  //   expect(withRootOrgDomain(uppercase)).not.toMatchSchema(schema);
  //   expect(withRootOrgDomain(domain)).toMatchSchema(schema);
  //   expect(withRootOrgDomain(spaces)).not.toMatchSchema(schema);
  //   expect(withRootOrgDomain(specialCharacters1)).not.toMatchSchema(schema);
  //   expect(withRootOrgDomain(specialCharacters2)).not.toMatchSchema(schema);
  // });

  // it("should validate root CA domain prefix", () => {
  //   const withRootCADomainPrefix = (d: string) =>
  //     updatedBase((json: FabloConfigJson) => {
  //       json.rootOrg.ca.prefix = d;
  //     });
  //
  //   expect(withRootCADomainPrefix(lettersOnly)).toMatchSchema(schema);
  //   expect(withRootCADomainPrefix(lettersAndNumber)).toMatchSchema(schema);
  //   expect(withRootCADomainPrefix(uppercase)).not.toMatchSchema(schema);
  //   expect(withRootCADomainPrefix(domain)).toMatchSchema(schema);
  //   expect(withRootCADomainPrefix(spaces)).not.toMatchSchema(schema);
  //   expect(withRootCADomainPrefix(specialCharacters1)).not.toMatchSchema(schema);
  //   expect(withRootCADomainPrefix(specialCharacters2)).not.toMatchSchema(schema);
  // });

  it("should validate first orderer org domain prefix", () => {
    const withFirstOrdererDomainPrefix = (d: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].orderer !== undefined ? (json.orgs[0].orderer.prefix = d) : undefined;
      });

    expect(withFirstOrdererDomainPrefix(lettersOnly)).toMatchSchema(schema);
    expect(withFirstOrdererDomainPrefix(lettersAndNumber)).toMatchSchema(schema);
    expect(withFirstOrdererDomainPrefix(domain)).toMatchSchema(schema);
    expect(withFirstOrdererDomainPrefix(uppercase)).not.toMatchSchema(schema);
    expect(withFirstOrdererDomainPrefix(spaces)).not.toMatchSchema(schema);
    expect(withFirstOrdererDomainPrefix(specialCharacters1)).not.toMatchSchema(schema);
    expect(withFirstOrdererDomainPrefix(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate root orderer consensus type ", () => {
    const withFirstOrdererConsensus = (c: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].orderer !== undefined ? (json.orgs[0].orderer.type = c as "solo" | "raft") : undefined;
      });

    expect(withFirstOrdererConsensus("solo")).toMatchSchema(schema);
    expect(withFirstOrdererConsensus("raft")).toMatchSchema(schema);
    expect(withFirstOrdererConsensus("kafka")).not.toMatchSchema(schema);
    expect(withFirstOrdererConsensus(lettersOnly)).not.toMatchSchema(schema);
  });

  it("should validate root orderer number of instances", () => {
    const withFirstOrdererNoOfInstances = (i: number) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].orderer !== undefined ? (json.orgs[0].orderer.instances = i) : undefined;
      });

    expect(withFirstOrdererNoOfInstances(1)).toMatchSchema(schema);
    expect(withFirstOrdererNoOfInstances(9)).toMatchSchema(schema);
    expect(withFirstOrdererNoOfInstances(0)).not.toMatchSchema(schema);
    expect(withFirstOrdererNoOfInstances(11)).not.toMatchSchema(schema);
  });

  it("should validate org name", () => {
    const withOrgName = (n: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].organization.name = n;
      });

    expect(withOrgName(lettersOnly)).toMatchSchema(schema);
    expect(withOrgName(lettersAndNumber)).toMatchSchema(schema);
    expect(withOrgName(uppercase)).toMatchSchema(schema);
    expect(withOrgName(domain)).not.toMatchSchema(schema);
    expect(withOrgName(spaces)).not.toMatchSchema(schema);
    expect(withOrgName(specialCharacters1)).not.toMatchSchema(schema);
    expect(withOrgName(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate org MSP name", () => {
    const withOrgMSPName = (n: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].organization.mspName = n;
      });

    expect(withOrgMSPName(lettersOnly)).toMatchSchema(schema);
    expect(withOrgMSPName(lettersAndNumber)).toMatchSchema(schema);
    expect(withOrgMSPName(uppercase)).toMatchSchema(schema);
    expect(withOrgMSPName(domain)).not.toMatchSchema(schema);
    expect(withOrgMSPName(spaces)).not.toMatchSchema(schema);
    expect(withOrgMSPName(specialCharacters1)).not.toMatchSchema(schema);
    expect(withOrgMSPName(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate org domain", () => {
    const withOrgDomain = (d: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].organization.domain = d;
      });

    expect(withOrgDomain(lettersOnly)).toMatchSchema(schema);
    expect(withOrgDomain(lettersAndNumber)).toMatchSchema(schema);
    expect(withOrgDomain(uppercase)).not.toMatchSchema(schema);
    expect(withOrgDomain(domain)).toMatchSchema(schema);
    expect(withOrgDomain(spaces)).not.toMatchSchema(schema);
    expect(withOrgDomain(specialCharacters1)).not.toMatchSchema(schema);
    expect(withOrgDomain(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate ca domain prefix", () => {
    const withCADomainPrefix = (d: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].ca.prefix = d;
      });

    expect(withCADomainPrefix(lettersOnly)).toMatchSchema(schema);
    expect(withCADomainPrefix(lettersAndNumber)).toMatchSchema(schema);
    expect(withCADomainPrefix(uppercase)).not.toMatchSchema(schema);
    expect(withCADomainPrefix(domain)).toMatchSchema(schema);
    expect(withCADomainPrefix(spaces)).not.toMatchSchema(schema);
    expect(withCADomainPrefix(specialCharacters1)).not.toMatchSchema(schema);
    expect(withCADomainPrefix(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate peer domain prefix", () => {
    const withPeerDomainPrefix = (d: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].peer.prefix = d;
      });

    expect(withPeerDomainPrefix(lettersOnly)).toMatchSchema(schema);
    expect(withPeerDomainPrefix(lettersAndNumber)).toMatchSchema(schema);
    expect(withPeerDomainPrefix(uppercase)).not.toMatchSchema(schema);
    expect(withPeerDomainPrefix(domain)).toMatchSchema(schema);
    expect(withPeerDomainPrefix(spaces)).not.toMatchSchema(schema);
    expect(withPeerDomainPrefix(specialCharacters1)).not.toMatchSchema(schema);
    expect(withPeerDomainPrefix(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate peer number of instances", () => {
    const withPeerNoOfInstances = (i: number) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].peer.instances = i;
      });

    expect(withPeerNoOfInstances(1)).toMatchSchema(schema);
    expect(withPeerNoOfInstances(9)).toMatchSchema(schema);
    expect(withPeerNoOfInstances(0)).not.toMatchSchema(schema);
    expect(withPeerNoOfInstances(101)).not.toMatchSchema(schema);
  });

  it("should validate peer database type", () => {
    const withPeerDatabaseType = (db: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].peer.db = db as "LevelDb" | "CouchDb";
      });

    expect(withPeerDatabaseType("LevelDb")).toMatchSchema(schema);
    expect(withPeerDatabaseType("CouchDb")).toMatchSchema(schema);
    expect(withPeerDatabaseType("MongoDb")).not.toMatchSchema(schema);
  });

  it("should validate peer tools", () => {
    const withTools = (tools: Record<string, unknown>) =>
      updatedBase((json: FabloConfigJson) => {
        json.orgs[0].tools = tools;
      });

    expect(withTools({ fabloRest: true })).toMatchSchema(schema);
    expect(withTools({ fabloRest: false })).toMatchSchema(schema);
    expect(withTools({})).toMatchSchema(schema);
    expect(withTools({ unknown: true })).not.toMatchSchema(schema);
  });

  it("should validate channel name - no spaces and capital letters", () => {
    const withChannelName = (n: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.channels[0].name = n;
      });

    expect(withChannelName(lettersOnly)).toMatchSchema(schema);
    expect(withChannelName(lettersAndNumber)).toMatchSchema(schema);
    expect(withChannelName(uppercase)).not.toMatchSchema(schema); // no uppercase in channel name
    expect(withChannelName(domain)).not.toMatchSchema(schema);
    expect(withChannelName(spaces)).not.toMatchSchema(schema);
    expect(withChannelName(specialCharacters1)).not.toMatchSchema(schema);
    expect(withChannelName(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate channel orderer org name", () => {
    const withChannelOrdererGroupName = (n: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.channels[0].ordererGroup = n;
      });

    expect(withChannelOrdererGroupName(lettersOnly)).toMatchSchema(schema);
    expect(withChannelOrdererGroupName(lettersAndNumber)).toMatchSchema(schema);
    expect(withChannelOrdererGroupName(uppercase)).toMatchSchema(schema);
    expect(withChannelOrdererGroupName(domain)).not.toMatchSchema(schema);
    expect(withChannelOrdererGroupName(spaces)).not.toMatchSchema(schema);
    expect(withChannelOrdererGroupName(specialCharacters1)).not.toMatchSchema(schema);
    expect(withChannelOrdererGroupName(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate chaincode name", () => {
    const withChaincodeName = (n: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.chaincodes[0].name = n;
      });

    expect(withChaincodeName(lettersOnly)).toMatchSchema(schema);
    expect(withChaincodeName(lettersAndNumber)).toMatchSchema(schema);
    expect(withChaincodeName(uppercase)).toMatchSchema(schema);
    expect(withChaincodeName(domain)).not.toMatchSchema(schema);
    expect(withChaincodeName(spaces)).not.toMatchSchema(schema);
    expect(withChaincodeName(specialCharacters1)).not.toMatchSchema(schema);
    expect(withChaincodeName(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate chaincode version", () => {
    const withChaincodeVersion = (n: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.chaincodes[0].version = n;
      });

    expect(withChaincodeVersion(lettersOnly)).toMatchSchema(schema);
    expect(withChaincodeVersion(lettersAndNumber)).toMatchSchema(schema);
    expect(withChaincodeVersion(uppercase)).toMatchSchema(schema);
    expect(withChaincodeVersion(domain)).toMatchSchema(schema);
    expect(withChaincodeVersion(spaces)).not.toMatchSchema(schema);
    expect(withChaincodeVersion(specialCharacters1)).not.toMatchSchema(schema);
    expect(withChaincodeVersion(specialCharacters2)).not.toMatchSchema(schema);
  });

  it("should validate chaincode language", () => {
    const withChaincodeLanguage = (l: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.chaincodes[0].lang = l as "java" | "golang" | "node";
      });

    expect(withChaincodeLanguage("java")).toMatchSchema(schema);
    expect(withChaincodeLanguage("node")).toMatchSchema(schema);
    expect(withChaincodeLanguage("golang")).toMatchSchema(schema);
    expect(withChaincodeLanguage("cobol")).not.toMatchSchema(schema);
  });

  it("should validate chaincode initialization arguments", () => {
    const withChaincodeInitialization = (i: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.chaincodes[0].init = i;
      });

    expect(withChaincodeInitialization(lettersOnly)).toMatchSchema(schema);
    expect(withChaincodeInitialization(lettersAndNumber)).toMatchSchema(schema);
    expect(withChaincodeInitialization(uppercase)).toMatchSchema(schema);
    expect(withChaincodeInitialization(domain)).toMatchSchema(schema);
    expect(withChaincodeInitialization(spaces)).toMatchSchema(schema);
    expect(withChaincodeInitialization(specialCharacters1)).toMatchSchema(schema);
    expect(withChaincodeInitialization(specialCharacters2)).toMatchSchema(schema);
  });

  it("should validate chaincode endorsement configuration", () => {
    const withChaincodeName = (n: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.chaincodes[0].endorsement = n;
      });

    expect(withChaincodeName(lettersOnly)).toMatchSchema(schema);
    expect(withChaincodeName(lettersAndNumber)).toMatchSchema(schema);
    expect(withChaincodeName(uppercase)).toMatchSchema(schema);
    expect(withChaincodeName(domain)).toMatchSchema(schema);
    expect(withChaincodeName(spaces)).toMatchSchema(schema);
    expect(withChaincodeName(specialCharacters1)).toMatchSchema(schema);
    expect(withChaincodeName(specialCharacters2)).toMatchSchema(schema);
  });

  it("should validate chaincode directory", () => {
    const withChaincodeName = (n: string) =>
      updatedBase((json: FabloConfigJson) => {
        json.chaincodes[0].directory = n;
      });

    expect(withChaincodeName(lettersOnly)).toMatchSchema(schema);
    expect(withChaincodeName(lettersAndNumber)).toMatchSchema(schema);
    expect(withChaincodeName(uppercase)).toMatchSchema(schema);
    expect(withChaincodeName(domain)).toMatchSchema(schema);
    expect(withChaincodeName(spaces)).toMatchSchema(schema);
    expect(withChaincodeName(specialCharacters1)).toMatchSchema(schema);
    expect(withChaincodeName(specialCharacters2)).toMatchSchema(schema);
  });

  it("should validate chaincode private data", () => {
    const withNoPrivateData = () =>
      updatedBase((json: FabloConfigJson) => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        json.chaincodes[0] = { ...json.chaincodes[0], privateData: undefined };
      });
    const withPrivateData = (d: PrivateDataJson[]) =>
      updatedBase((json: FabloConfigJson) => {
        json.chaincodes[0].privateData = d;
      });
    const privateData = (name: string, ...orgNames: string[]) => ({ name, orgNames });
    const validOrgName = base.orgs[0].organization.name;
    const withPrivateDataName = (name: string) => withPrivateData([privateData(name, validOrgName)]);

    // various names
    expect(withPrivateDataName(lettersOnly)).toMatchSchema(schema);
    expect(withPrivateDataName(underscore)).toMatchSchema(schema);
    expect(withPrivateDataName(lettersAndNumber)).toMatchSchema(schema);
    expect(withPrivateDataName(uppercase)).toMatchSchema(schema);
    expect(withPrivateDataName(domain)).not.toMatchSchema(schema);
    expect(withPrivateDataName(spaces)).not.toMatchSchema(schema);
    expect(withPrivateDataName(specialCharacters1)).not.toMatchSchema(schema);
    expect(withPrivateDataName(specialCharacters2)).not.toMatchSchema(schema);

    // no private data, wrong object, two objects
    expect(withNoPrivateData()).toMatchSchema(schema);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    expect(withPrivateData({ wrong: "obj" })).not.toMatchSchema(schema);
    expect(
      withPrivateData([privateData(lettersAndNumber, validOrgName), privateData(lettersOnly, validOrgName)]),
    ).toMatchSchema(schema);
  });
});
