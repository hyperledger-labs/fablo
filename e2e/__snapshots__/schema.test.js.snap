// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema should match snapshot 1`] = `
Object {
  "$id": "http://example.com/example.json",
  "$schema": "http://json-schema.org/draft-07/schema",
  "additionalProperties": false,
  "default": Object {},
  "properties": Object {
    "$schema": Object {
      "const": "https://github.com/softwaremill/fabrica/releases/download/0.1.0-unstable/schema.json",
    },
    "chaincodes": Object {
      "$id": "#/properties/chaincodes",
      "default": Array [
        Object {
          "channel": "channel1",
          "directory": "chaincode1",
          "endorsement": "AND ('Org1.member')",
          "init": "{'Args':[]}",
          "lang": "java",
          "name": "chaincode1",
          "version": "0.0.1",
        },
      ],
      "items": Object {
        "$id": "#/properties/chaincodes/items",
        "additionalProperties": false,
        "properties": Object {
          "channel": Object {
            "$id": "#/properties/chaincodes/items/properties/channel",
            "enumSource": Array [
              Array [],
              Object {
                "source": "channels",
                "value": "{{item.key}}",
              },
            ],
            "pattern": "^[a-z0-9]+$",
            "title": "Channel key",
            "type": "string",
            "watch": Object {
              "channels": "channels",
            },
          },
          "directory": Object {
            "$id": "#/properties/chaincodes/items/properties/directory",
            "title": "Chaincode directory",
            "type": "string",
          },
          "endorsement": Object {
            "$id": "#/properties/chaincodes/items/properties/endorsement",
            "title": "Endorsement configuration",
            "type": "string",
          },
          "init": Object {
            "$id": "#/properties/chaincodes/items/properties/init",
            "title": "Initialization arguments",
            "type": "string",
          },
          "lang": Object {
            "$id": "#/properties/chaincodes/items/properties/lang",
            "enum": Array [
              "golang",
              "java",
              "node",
            ],
            "title": "Language",
            "type": "string",
          },
          "name": Object {
            "$id": "#/properties/chaincodes/items/properties/name",
            "pattern": "^[a-zA-Z0-9_-]+$",
            "title": "Name",
            "type": "string",
          },
          "privateData": Object {
            "$id": "#/properties/chaincodes/items/properties/privateData",
            "items": Object {
              "$id": "#/properties/chaincodes/items/properties/privateData/items",
              "additionalProperties": false,
              "properties": Object {
                "name": Object {
                  "$id": "#/properties/chaincodes/items/properties/privateData/items/properties/name",
                  "pattern": "^[A-Za-z0-9_-]+$",
                  "title": "Collection name",
                  "type": "string",
                },
                "orgNames": Object {
                  "$id": "#/properties/chaincodes/items/properties/privateData/items/properties/orgNames",
                  "items": Object {
                    "$id": "#/properties/chaincodes/items/properties/privateData/items/properties/orgNames/items",
                    "enumSource": Array [
                      Array [],
                      Object {
                        "source": "orgs",
                        "value": "{{item.organization.name}}",
                      },
                    ],
                    "pattern": "^[A-Za-z0-9]+$",
                    "title": "Organization name",
                    "type": "string",
                    "watch": Object {
                      "orgs": "orgs",
                    },
                  },
                  "title": "Organization names",
                  "type": "array",
                },
              },
              "required": Array [
                "name",
                "orgNames",
              ],
              "title": "Private data collection",
              "type": "object",
            },
            "title": "Private data collections",
            "type": "array",
          },
          "version": Object {
            "$id": "#/properties/chaincodes/items/properties/version",
            "pattern": "^[a-zA-Z0-9\\\\.]+$",
            "title": "Version",
            "type": "string",
          },
        },
        "required": Array [
          "name",
          "version",
          "lang",
          "channel",
          "init",
          "endorsement",
          "directory",
        ],
        "title": "Chaincode",
        "type": "object",
      },
      "title": "Chaincodes",
      "type": "array",
    },
    "channels": Object {
      "$id": "#/properties/channels",
      "default": Array [
        Object {
          "key": "channel1",
          "name": "my-channel1",
          "orgs": Array [
            Object {
              "key": "org1",
              "peers": Array [
                "peer0",
              ],
            },
          ],
        },
        Object {
          "key": "channel2",
          "name": "my-channel2",
          "orgs": Array [
            Object {
              "key": "org1",
              "peers": Array [
                "peer1",
              ],
            },
          ],
        },
      ],
      "items": Object {
        "$id": "#/properties/channels/items",
        "additionalProperties": false,
        "properties": Object {
          "key": Object {
            "$id": "#/properties/channels/items/properties/key",
            "pattern": "^[a-z0-9]+$",
            "title": "Key",
            "type": "string",
          },
          "name": Object {
            "$id": "#/properties/channels/items/properties/name",
            "pattern": "^[a-z0-9\\\\-]+$",
            "title": "Name",
            "type": "string",
          },
          "orgs": Object {
            "$id": "#/properties/channels/items/properties/orgs",
            "items": Object {
              "$id": "#/properties/channels/items/properties/orgs/items",
              "additionalProperties": false,
              "properties": Object {
                "key": Object {
                  "$id": "#/properties/channels/items/properties/orgs/items/properties/key",
                  "enumSource": Array [
                    Array [],
                    Object {
                      "source": "orgs",
                      "value": "{{item.organization.key}}",
                    },
                  ],
                  "pattern": "^[a-z0-9]+$",
                  "title": "Organization key",
                  "type": "string",
                  "watch": Object {
                    "orgs": "orgs",
                  },
                },
                "peers": Object {
                  "$id": "#/properties/channels/items/properties/orgs/items/properties/peers",
                  "items": Object {
                    "$id": "#/properties/channels/items/properties/orgs/items/properties/peers/items",
                    "pattern": "^[a-z0-9]+$",
                    "title": "Peer",
                    "type": "string",
                  },
                  "title": "Peers for ogranization",
                  "type": "array",
                },
              },
              "required": Array [
                "key",
                "peers",
              ],
              "title": "Organization",
              "type": "object",
            },
            "title": "Organizations in the channel",
            "type": "array",
          },
        },
        "required": Array [
          "key",
          "name",
          "orgs",
        ],
        "title": "Channel",
        "type": "object",
      },
      "title": "Channels",
      "type": "array",
    },
    "networkSettings": Object {
      "$id": "#/properties/networkSettings",
      "additionalProperties": false,
      "description": "Basic settings of the Hyperledger Fabric network",
      "properties": Object {
        "fabricVersion": Object {
          "$id": "#/properties/networkSettings/properties/fabricVersion",
          "default": "1.4.6",
          "enum": Array [
            "1.3.0",
            "1.4.0",
            "1.4.1",
            "1.4.2",
            "1.4.3",
            "1.4.4",
            "1.4.5",
            "1.4.6",
            "1.4.7",
            "1.4.8",
            "1.4.9",
            "1.4.10",
            "1.4.11",
          ],
          "title": "Hyperledger Fabric version",
          "type": "string",
        },
        "monitoring": Object {
          "$id": "#/properties/monitoring",
          "additionalProperties": false,
          "description": "Optional settings for monitoring purposes",
          "properties": Object {
            "loglevel": Object {
              "$id": "#/properties/monitoring/properties/loglevel",
              "default": "info",
              "enum": Array [
                "debug",
                "info",
                "warn",
              ],
              "title": "Log level for all components",
              "type": "string",
            },
          },
          "required": Array [],
          "title": "Monitoring settings",
          "type": "object",
        },
        "tls": Object {
          "$id": "#/properties/networkSettings/properties/tls",
          "default": true,
          "title": "Use TLS",
          "type": "boolean",
        },
      },
      "required": Array [
        "fabricVersion",
        "tls",
      ],
      "title": "Network settings",
      "type": "object",
    },
    "orgs": Object {
      "$id": "#/properties/orgs",
      "default": Array [
        Object {
          "ca": Object {
            "prefix": "ca",
          },
          "organization": Object {
            "domain": "org1.com",
            "key": "org1",
            "mspName": "Org1MSP",
            "name": "Org1",
          },
          "peer": Object {
            "db": "LevelDb",
            "instances": 2,
            "prefix": "peer",
          },
        },
      ],
      "items": Object {
        "$id": "#/properties/orgs/items",
        "additionalProperties": false,
        "properties": Object {
          "ca": Object {
            "$id": "#/properties/orgs/items/properties/ca",
            "additionalProperties": false,
            "properties": Object {
              "prefix": Object {
                "$id": "#/properties/orgs/items/properties/ca/properties/prefix",
                "pattern": "^[a-z0-9\\\\.\\\\-]+$",
                "title": "Domain prefix",
                "type": "string",
              },
            },
            "required": Array [
              "prefix",
            ],
            "title": "Organization Certificate Authority (CA)",
            "type": "object",
          },
          "organization": Object {
            "$id": "#/properties/orgs/items/properties/organization",
            "additionalProperties": false,
            "properties": Object {
              "domain": Object {
                "$id": "#/properties/orgs/items/properties/organization/properties/domain",
                "pattern": "^[a-z0-9\\\\.\\\\-]+$",
                "title": "Domain",
                "type": "string",
              },
              "key": Object {
                "$id": "#/properties/orgs/items/properties/organization/properties/key",
                "pattern": "^[a-z0-9]+$",
                "title": "Key",
                "type": "string",
              },
              "mspName": Object {
                "$id": "#/properties/orgs/items/properties/organization/properties/mspName",
                "pattern": "^[a-zA-Z0-9]+$",
                "title": "MSP Name",
                "type": "string",
              },
              "name": Object {
                "$id": "#/properties/orgs/items/properties/organization/properties/name",
                "pattern": "^[a-zA-Z0-9]+$",
                "title": "Name",
                "type": "string",
              },
            },
            "required": Array [
              "key",
              "name",
              "mspName",
              "domain",
            ],
            "title": "Basic information",
            "type": "object",
          },
          "peer": Object {
            "$id": "#/properties/orgs/items/properties/peer",
            "additionalProperties": false,
            "properties": Object {
              "anchorPeerInstances": Object {
                "$id": "#/properties/orgs/items/properties/peer/properties/anchorPeerInstances",
                "maximum": 100,
                "minimum": 1,
                "title": "Number of anchor peer instances",
                "type": "integer",
              },
              "db": Object {
                "$id": "#/properties/orgs/items/properties/peer/properties/db",
                "enum": Array [
                  "LevelDb",
                  "CouchDb",
                ],
                "title": "Database Type",
                "type": "string",
              },
              "instances": Object {
                "$id": "#/properties/orgs/items/properties/peer/properties/instances",
                "maximum": 100,
                "minimum": 1,
                "title": "Number of instances",
                "type": "integer",
              },
              "prefix": Object {
                "$id": "#/properties/orgs/items/properties/peer/properties/prefix",
                "pattern": "^[a-z0-9\\\\.\\\\-]+$",
                "title": "Domain prefix",
                "type": "string",
              },
            },
            "required": Array [
              "prefix",
              "instances",
              "db",
            ],
            "title": "Peer",
            "type": "object",
          },
        },
        "required": Array [
          "organization",
          "ca",
          "peer",
        ],
        "title": "Organization",
        "type": "object",
      },
      "title": "Organizations",
      "type": "array",
    },
    "rootOrg": Object {
      "$id": "#/properties/rootOrg",
      "additionalProperties": false,
      "description": "The configuration of the root organization for the Hyperledger Fabric network.",
      "properties": Object {
        "ca": Object {
          "$id": "#/properties/rootOrg/properties/ca",
          "additionalProperties": false,
          "default": Object {
            "prefix": "ca",
          },
          "properties": Object {
            "prefix": Object {
              "$id": "#/properties/rootOrg/properties/ca/properties/prefix",
              "pattern": "^[a-z0-9\\\\.\\\\-]+$",
              "title": "Domain prefix",
              "type": "string",
            },
          },
          "required": Array [
            "prefix",
          ],
          "title": "Root Certificate Authority (CA)",
          "type": "object",
        },
        "orderer": Object {
          "$id": "#/properties/rootOrg/properties/orderer",
          "additionalProperties": false,
          "default": Object {
            "instances": 1,
            "prefix": "orderer",
            "type": "solo",
          },
          "properties": Object {
            "instances": Object {
              "$id": "#/properties/rootOrg/properties/orderer/properties/instances",
              "maximum": 10,
              "minimum": 1,
              "title": "Number of instances",
              "type": "integer",
            },
            "prefix": Object {
              "$id": "#/properties/rootOrg/properties/orderer/properties/prefix",
              "pattern": "^[a-z0-9\\\\.\\\\-]+$",
              "title": "Domain prefix",
              "type": "string",
            },
            "type": Object {
              "$id": "#/properties/rootOrg/properties/orderer/properties/type",
              "description": "The 'solo' consensus type may be used in development environment only. Use 'raft' in production.",
              "enum": Array [
                "solo",
                "raft",
              ],
              "title": "Orderer type",
              "type": "string",
            },
          },
          "required": Array [
            "prefix",
            "type",
            "instances",
          ],
          "title": "Orderer",
          "type": "object",
        },
        "organization": Object {
          "$id": "#/properties/rootOrg/properties/organization",
          "additionalProperties": false,
          "default": Object {
            "domain": "root.com",
            "key": "root",
            "mspName": "OrdererMSP",
            "name": "Orderer",
          },
          "properties": Object {
            "domain": Object {
              "$id": "#/properties/rootOrg/properties/organization/properties/domain",
              "pattern": "^[a-z0-9\\\\.\\\\-]+$",
              "title": "Domain",
              "type": "string",
            },
            "key": Object {
              "$id": "#/properties/rootOrg/properties/organization/properties/key",
              "pattern": "^[a-z0-9]+$",
              "title": "Key",
              "type": "string",
            },
            "mspName": Object {
              "$id": "#/properties/rootOrg/properties/organization/properties/mspName",
              "description": "Membership Service Provider (MSP) is a component that offers an abstraction of membership operations.",
              "pattern": "^[a-zA-Z0-9]+$",
              "title": "MSP Name",
              "type": "string",
            },
            "name": Object {
              "$id": "#/properties/rootOrg/properties/organization/properties/name",
              "pattern": "^[a-zA-Z0-9]+$",
              "title": "Name",
              "type": "string",
            },
          },
          "required": Array [
            "key",
            "name",
            "mspName",
            "domain",
          ],
          "title": "Basic information",
          "type": "object",
        },
      },
      "required": Array [
        "organization",
        "ca",
        "orderer",
      ],
      "title": "Root organization",
      "type": "object",
    },
  },
  "required": Array [
    "networkSettings",
    "rootOrg",
    "orgs",
    "channels",
    "chaincodes",
  ],
  "title": "Root",
  "type": "object",
}
`;
